docs/madr/decisions/031-locality-aware-load-balancing.md

```diff
@@ -0,0 +1,292 @@
+
+#### Matching endpoints
+Each instance of the service might be in a different locality: zone, node. Users need to properly tag these dataplanes so we are able to match them. We need to retrieve all tags from all inbounds and, later, based on these tags, match outbounds that are in the same location.
+
+Example:
+
+On Kubernetes, all tags are populated automatically, while on Universal, users need to populate tags themselves:
+
+```yaml
+type: Dataplane
+mesh: default
+name: demo
+networking:
+  address: 127.0.0.1
+  inbound:
+    - port: 9999
+      servicePort: 10000
+      serviceAddress: 127.0.0.1
+      tags:
+        kuma.io/service: demo
+        kuma.io/protocol: tcp
+        k8s.io/az: us-east-1
+```
+
+
+Later, based on these tags, the control-plane creates `ClusterLoadBalancing` and puts all tags, for each endpoint, into cluster metadata::
+
+```json
+"metadata":{
+  "filter_metadata": {
+    "envoy.lb": {
+      "az/test": "b",
+      "kuma.io/protocol": "tcp",
+      "k8s.io/az": "us-east-1"
+    }
+  }
+}
+```
+
+
+We can retrieve them from `ClusterLoadBalancing` and match them with the policy. First, we need to retrieve all tags from inbounds so we can determine our location and match it with endpoints.
+
+Pseudo algorithm:
+
+1. Get all inbound tags.
+2. Take the matched rules.
+3. Check if inbound has the tag defined in the rule. 
+4. If yes, iterate over all endpoints of the service and group them in to priority groups, by matching with the specific rule( first inZone, then  cross Zone). If no, get all the endpoints in the zone. 
+5. Create `LocalityLbEndpoints` with selected endpoints. 
+6. Override `ClusterLoadAssignment` for the dataplane in the `ResourceStore`. 
+
+Dataplanes in the specific location are going to receive only endpoints in the specific locality. For cross zone traffic without egress the control-plane delivers only ingresses of specific zones.
+
+##### Egress
+Egress is not as simple as ingress. Currently, we support Locality Aware when atleast one client requires it, so we cannot distinguish between clients. Control-plane needs to send all dataplanes to the egress sidecar because there might be services sending requests to all zones. There are 2 options:
+- LoadBalacing based on metadata
+- Separate clusters
+
+Load balancing based on the metadata seems to be able to solve an issue but it has limitations. Because, egress knows all endpoints and does routing based on matching metadata the request might not obey the priority of zones. In this case we cannot configure different priorities for different clients.
+
+Separate cluster seems to be the only option that could solve this problem. When there is a different configuration of Load Balancing for the one of dataplane, control-plane sends to egress new a cluster `mesh-name:service-name:hash(tag + value)` with the routing based on the SNI: `service-name{mesh=mesh-name,hash=hash}`
```

Comment:
I'm not sure I understand how we want to generate separate clusters. Is it a cluster per client or per destination?

Reply:
Cluster per client. Each client might define different priorities for cross zone communication. 
Servica A [zone-1 : 1, zone-2: 2]
Service B [zone2: 1]
In the case of egress, it knows about all zones so if we are going to use one cluster we can't ensure that traffic is routed based on the provided definition for the client.

Reply:
What about SNI? IIRC ZoneEgress doesn't rewrite SNI at this moment and ZoneIngress later is using the same SNI in filterChainMatches. 
1. Is it possible to rewrite SNI without TLS termination?
2. If it's not possible, does it mean ZoneIngress has to generate filterChainMatches using clients' hashes?

Reply:
1. No, it isn't
2. Yes, ingresses would need filter chains with hashes to make it work.

Reply:
> Is it possible to rewrite SNI without TLS termination?
Why do we need to rewrite SNI?

Reply:
1. Egress has cluster definition `service-a`, when `client-1` wants to make a request to `service-a` but only in `zone-1` and `zone-2` but client `client-2` wants to make a request to `service -a` in `zone-3` and `zone-4` with lower priority egress won't use client logis but global one.
2. Because when we override SNI ingress needs to match SNI from the request and in this case it might not be the same

Reply:
I'm worried about the scalability of this approach. Imagine a bit extreme but still possible use case when a user creates MLBS for each service:
```yaml
targetRef:
  kind: MeshService
  name: "${service}"
to:
  targetRef:
    kind: Mesh
  defaults:
    localityAwareness: {...}
```

In that case, the number of filterChainMatches in ZoneIngress is `num_services*num_of_destinations`, while today it's `num_of_destinations`

Reply:
Some thoughts that I shared with @lukidzi thinking out loud on priv.

Keep in mind that AFAIU for crosszone routing we only care about routing by zone.

I also don't think separate clusters on egress are the way to go.
Let's say we have a service in zone1 (my local), zone2 zone3, zone4 and we only want to fallback to zone2+zone3.

It's easy with Ingress, because we set priorities
local endpoint - zone1: 0
zone ingress of zone2: 1
zone ingress of zone3: 1

It's hard with Egress, because all zone ingresses are hidden behind one zone egress.
If we solve the problem that we can put multiple endpoints with different metadata we could do this:
local endpoint - zone1: 0
zone egress: 1 that sends SNI: `serviceA{mesh=default,kuma.io/zone=zone2}`
zone egress: 1 that sends SNI: `serviceA{mesh=default,kuma.io/zone=zone3}`

At this moment we cannot do this because Envoy limitation, but it's not the first time it's a problem for us. Maybe we could solve this.

Alternatively, we could consider separate clusters but on the client side and try to use https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/aggregate_cluster

Reply:
I've included @jakubdyszkiewicz's suggestion with all the details. I feel like it's possible but need to confirm the aggregate clusters part.